File: Python/Depth-First Search/17. Letter Combinations of a Phone Number.py
from typing import List

class Solution:
    lookup = ['', '', ['a', 'b', 'c'], ['d', 'e', 'f'], ['g', 'h', 'i'], 
              ['j', 'k', 'l'], ['m', 'n', 'o'], ['p', 'q', 'r', 's'], ['t', 'u', 'v'], ['w', 'x', 'y', 'z']]
    
    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0: return []
        firstD = self.lookup[ord(digits[0]) - ord('0')]
        if len(digits) == 1: return firstD
        res = []
        restD = self.letterCombinations(digits[1:])
        for c1 in firstD:
            for c2 in restD:
                res.append(c1 + c2)
        return res
        

File: C++/Depth-First Search/17. Letter Combinations of a Phone Number.cpp
class Solution {
    
    vector<vector<string>> lookup = {{},{},{"a","b","c"},{"d","e","f"},{"g","h","i"},
    {"j","k","l"},{"m","n","o"},{"p","q","r","s"},{"t","u","v"},{"w","x","y","z"}};
    
public:
    vector<string> letterCombinations(string digits) {
        if (digits.size() == 0) return {};
        vector<string> firstD = lookup[digits[0] - '0'];
        if (digits.size() == 1) return firstD;
        vector<string> res;
        // substr(startPos, len)
        vector<string> restD = letterCombinations(digits.substr(1));
        for (string& c1: firstD) {
            for (string& c2: restD) {
                res.emplace_back(c1+c2);
            }
        }
        return res;
    }
};